bb.1
.TH BB 1 "December 16, 2001"
.SH NAME
bb \- an ASCII-art demo
.SH SYNOPSIS
.B bb
.RB [ options ]
.SH DESCRIPTION
.B bb
is a high quality audio-visual demonstration for your text terminal.
.SH OPTIONS
A summary of options are included below.
For a complete description type 
.BR bb 
.BR -help .
.TP
.B \-help
Show summary of options.
.TP
.B \-loop
Play demo in infinite loop.
.TP
.B \-driver
Select driver. Available drivers: linux, slang, X11, stdout, stderr.
.TP
.B \-kbddriver
Select keyboard driver. Available drivers: slang, X11, stdin.
.TP
.B \-mousedriver
Select mouse driver. Available drivers: X11, gpm, dos.
.TP
.B \-width
Set width.
.TP
.B \-height
Set height.
.TP
.B \-minwidth
Set minimum width.
.TP
.B \-minheight
Set minimum height.
.TP
.B \-maxwidth
Set maximum width.
.TP
.B \-maxheight
Set maximum height.
.TP
.B \-recwidth
Set recommended width.
.TP
.B \-recheight
Set recommended height.
.TP
.B \-dim
Enable usage of dim (half bright) attribute.
.TP
.B \-bold
Enable usage of bold (double bright) attribute.
.TP
.B \-reverse
Enable usage of reverse attribute.
.TP
.B \-normal
Enable usage of normal attribute.
.TP
.B \-boldfont
Enable usage of boldfont attribute.
.TP
.B \-no<attr>
Disable attribute (i.e
.BR -nobold ).
.TP
.B \-extended
Use all 256 characters.
.TP
.B \-eight
Use eight bit ASCII.
.TP
.B \-font <font>
Select font. This option is used on hardware where aalib is unable to
determine the current font. Available fonts: vga8, vga9, mda14, vga14, X8x13,
X8x16, X8x13bold, vgagl8, line.
.TP
.B \-inverse
Enable inverse rendering.
.TP
.B \-noinverse
Disable inverse rendering.
.TP
.B \-bright <val>
Set brightness (0-255).
.TP
.B \-contrast <val>
Set contrast (0-255).
.TP
.B \-gamma <val>
Set gamma correction value (0-1).
.TP
.B \-nodither
Disable dithering.
.TP
.B \-floyd_steinberg
Floyd-Steinberg dithering.
.TP
.B \-error_distribution
Error distribution dithering.
.TP
.B \-random <val>
Set random dithering value (0-inf).
.TP
.B \-dimmul <val>
Multiply factor for dim attribute (5.3).
.TP
.B \-boldmul <val>
Multiply factor for bold attribute (2.7).
.SH BUGS
Please report any bugs you find to Jan Hubicka <hubicka@freesoft.cz>.
.SH LICENCE
.B bb
is covered by the GNU General Public License (GPL).
.SH SEE ALSO
.BR aafire (1),
.BR aainfo (1),
.BR aasavefont (1),
.BR aatest (1),
.BR aview (1),
.BR asciiview (1),
.BR aaflip (1),
.BR aa3d (1),
.BR xaos (6),
.BR aatv (1).
.SH AUTHOR
Jan Hubicka <hubicka@freesoft.cz>
.PP
This manual page was written by Michael Bramer <grisu@debian.org>,
for the Debian GNU/Linux system (but may be used by others).


debian/changelog
bb (1.3rc1-9) UNRELEASED; urgency=medium

  * Add myself to Uploaders.
  * Apply "wrap-and-sort -a".
  * Revamp debian/rules:
    + Use dh_auto_{configure,build,install,clean}.
    + Remove now redundant removal of stamp files
    + Remove parsing of DEB_BUILD_OPTIONS, it's now done automatically.
    + Remove no more needed rpath mangling of generated Makefile.
    + Finally switch to a minimal dh v7 style debian/rules file.
  * Convert to source format "3.0 (quilt)" using diff2patches.
    + Assign changes to bug reports or changelog items where obvious.
  * Enabled all hardening build flags.
  * Add simple debian/watch file (no version mangling yet).
  * Support parallel building.
  * Declare compliance with Debian Policy 3.9.8. (No changes needed.)
  * Add patch against textform.c to fix typos found by lintian.
  * Convert debian/copyright to machine-readable DEP-5 format and update
    copyright years and owners.
  * debian/copyright: Update addresses of XaoS homepage and AAlib mailing
    lists. Issues reported by DUCK.

 -- Axel Beckert <abe@debian.org>  Sat, 23 Jul 2016 03:31:20 +0200

bb (1.3rc1-8.5) unstable; urgency=medium

  * Non-maintainer upload.

  [ Axel Beckert ]
  * Move from no more supported debhelper compatibility level 4 to
    currently recommended level 9 and make debian/rules and debian/control
    more future-proof:
    + Change "dh_clean -k" to "dh_prep".
    + Remove explicit argument from dh_installchangelogs. (No more needed.)
    + Add recommended targets build-indep and build-arch.
    + Add ${misc:Depends} to Depends
    + Closes: #817373
  * Remove config.guess and config.sub from Debian's .diff.gz and use
    dh_autotools-dev_{restore,update}config to provide up-to-date files at
    build time.
    + Add build-dependency on autotools-dev.

  [ Logan Rosen ]
  * debian/rules: Don't ignore all distclean errors.
  * debian/copyright: Don't point to symlink GPL.

 -- Axel Beckert <abe@debian.org>  Thu, 21 Jul 2016 19:21:48 +0200

bb (1.3rc1-8.4) unstable; urgency=medium

  * Non-maintainer upload.
  * Drop the dependency on oss-compat, it's no longer needed. libmikmod
    now supports ALSA. Closes: #617426.
  * Build-depend on libmikmod-dev instead of libmikmod2-dev.
    Closes: #744906.

 -- Stephen Kitt <skitt@debian.org>  Sun, 21 Jun 2015 20:30:55 +0200

bb (1.3rc1-8.3) unstable; urgency=medium

  * Non-maintainer upload.
  * Improve README.Debian from previous upload:
    + Mention how to run bb under PulseAudio anyways (Thanks Anthony
      DeRobertis!)
    + Fix typo (Thanks Antoine Amarilli!)

 -- Axel Beckert <abe@debian.org>  Sun, 12 Apr 2015 22:10:43 +0200

bb (1.3rc1-8.2) unstable; urgency=medium

  * Non-maintainer upload.
  * Mitigate hanging visuals in combination with PulseAudio.
    + Set default answer for Music to "no".
    + Add README.Debian explaining the situation.
    + Mitigates: #761023

 -- Axel Beckert <abe@debian.org>  Wed, 01 Apr 2015 21:35:03 +0200

bb (1.3rc1-8.1) unstable; urgency=low

  * Non-maintainer upload.
  * Makefile.am, Makefile.in: Added -lm to bb_LDADD.  Fixes FTBFS.
    Patch by Bhavani Shankar R <bhavi@ubuntu.com>.  Closes: #553954, #608869.
  * debian/control: Added Homepage field.  Closes: #495048.

 -- Bart Martens <bartm@debian.org>  Sun, 02 Oct 2011 15:16:31 +0200

bb (1.3rc1-8) unstable; urgency=low

  * Depend on oss-compat to make music/sound work again. This іs
    bug #385844 and should really be fixed in libmikmod2, but until that
    is done I'll fix it in bb locally.
  * Fix typo in description (Closes: #363176).
  * debian/copyright:
    + New FSF address (lintian).
    + Be more verbose, print license of all files.
    + My packaging is public domain.
  * Standards-Version: 3.7.2.2 (no changes required).

 -- Uwe Hermann <uwe@debian.org>  Fri,  2 Mar 2007 15:36:01 +0100

bb (1.3rc1-7) unstable; urgency=low

  * Fix FTBS with gcc 3.4 and 4.0 (Closes: #317415).
    Thanks to Karl Chen for the patch.
  * Build against libaa1 instead of aalib1 (Closes: #317167).
  * Standards-Version: 3.6.2.0 (no changes required).
  * Moved bb from /usr/bin to /usr/games (Closes: #260536).

 -- Uwe Hermann <uwe@debian.org>  Fri,  8 Jul 2005 19:55:42 +0200

bb (1.3rc1-6) unstable; urgency=low

  * Fixed segfault on amd64 and alpha (Closes: #304434, #32160).
    Thanks go to Corey Hickey, Kaare Hviid and Helge Kreutzmann for the
    patch and for testing it on amd64 and alpha.
  * Updated debian/copyright.

 -- Uwe Hermann <uwe@debian.org>  Mon, 25 Apr 2005 02:45:35 +0200

bb (1.3rc1-5) unstable; urgency=low

  * Improved description (Closes: #209430).
  * Standards-Version: 3.6.1.0 (no changes).
  * debian/menu: Fixed Typos.

 -- Uwe Hermann <uwe@debian.org>  Thu, 25 Sep 2003 20:19:45 +0200

bb (1.3rc1-4) unstable; urgency=low

  * debian/menu: s/needs=vc/needs=text/ (Closes: #195075).
  * Standards-Version: 3.5.9.0.
  * Build-Depends: debhelper (>> 4.0.0).
  * Use a recent dh_make template for debian/rules.

 -- Uwe Hermann <uwe@debian.org>  Tue,  3 Jun 2003 16:33:48 +0200

bb (1.3rc1-3) unstable; urgency=low

  * debian/changelog: Removed emacs cruft.
  * Changed my email address to uwe@debian.org.
  * debian/rules: Use a recent dh_make template.

 -- Uwe Hermann <uwe@debian.org>  Sun, 24 Mar 2002 23:53:40 +0100

bb (1.3rc1-2) unstable; urgency=low

  * Enable hack in backconv.c, so that Joey Hess <joey@kitenet.net> can
    remove a far uglier hack in aalib (Closes: #112039).
  * Build-Depend on aalib-dev >= 1.4p5-7 (the first version without the hack).
  * debian/menu: Use X11 and vc, not X11 and text.
  * bb.1: Fix "SEE ALSO" link. It's aa3d(1) not aa3d(6).

 -- Uwe Hermann <uh1763@hermann-uwe.de>  Sun, 30 Dec 2001 20:08:43 +0100

bb (1.3rc1-1) unstable; urgency=low

  * New maintainer (with permission from Michael Bramer <grisu@debian.org>).
  * Repackaged from scratch using debhelper v3.
  * Acknowledge NMU (Thanks Steve!).
    Closes: #84976, #101453, #103171, #107905, #117144, #112200.
  * bb.1:

bb.c
/*
 * BB: The portable demo
 *
 * (C) 1997 by AA-group (e-mail: aa@horac.ta.jcu.cz)
 *
 * 3rd August 1997
 * version: 1.2 [final3]
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public Licences as by published
 * by the Free Software Foundation; either version 2; or (at your option)
 * any later version
 *
 * This program is distributed in the hope that it will entertaining,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILTY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 * Publis License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.
 * 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <string.h>
#include <stdlib.h>
#include <ctype.h>

#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif

#include <aalib.h>
#include "bb.h"

int finish_stuff, starttime, endtime;
int dual = 0;
static int quitnow = 0;
int loopmode;
aa_context *context;
aa_renderparams *params;
int TIME;
tl_timer *scenetimer;
struct font *font;

double getwidth(double size)
{
    double height = aa_imgheight(context) / size;
    double width = height * (double) aa_imgwidth(context) * 0.75 / aa_imgheight(context) * aa_mmheight(context) / aa_mmwidth(context);
    return (width);
}

void centerprint(int x, int y, double size, int color, char *text, int mode)
{
    if (!dual || !mode) {
	double height = aa_imgheight(context) / size;
	double width = height * (double) aa_imgwidth(context) * 0.75 / aa_imgheight(context) * aa_mmheight(context) / aa_mmwidth(context);
	print(x - (width * strlen(text)) / 2, y - height / 2, width, height, font, color, text);
    }
    else {
	if (mode & 1) {
	    double height = aa_imgheight(context) / size;
	    double width = height * (double) aa_imgwidth(context) * 0.75 / aa_imgheight(context) * aa_mmheight(context) / aa_mmwidth(context);
	    print(x / 2 - (width * strlen(text)) / 2, y - height / 2, width, height, font, color, text);
	}
	if (mode & 2) {
	    double height = aa_imgheight(context) / size;
	    double width = height * (double) aa_imgwidth(context) * 0.75 / aa_imgheight(context) * aa_mmheight(context) / aa_mmwidth(context);
	    print(aa_imgwidth(context) / 2 + x / 2 - (width * strlen(text)) / 2, y - height / 2, width, height, font, color, text);
	}
    }
}

void centerprinth(int x, int y, double size, int color, char *text, int mode)
{
    if (!mode || !dual) {
	double width = aa_imgwidth(context) / size;
	double height = width * (double) aa_imgheight(context) * 1.333 / aa_imgwidth(context) * aa_mmwidth(context) / aa_mmheight(context);
	print(x - (width * strlen(text)) / 2, y - height / 2, width, height, font, color, text);
    }
    else {
	if (mode & 1) {
	    double width = aa_imgwidth(context) / size / 2;
	    double height = width * (double) aa_imgheight(context) * 1.333 / aa_imgwidth(context) * aa_mmwidth(context) / aa_mmheight(context);
	    print(x / 2 - (width * strlen(text)) / 2, y - height / 2, width, height, font, color, text);
	}
	if (mode & 1) {
	    double width = aa_imgwidth(context) / size / 2;
	    double height = width * (double) aa_imgheight(context) * 1.333 / aa_imgwidth(context) * aa_mmwidth(context) / aa_mmheight(context);
	    print(aa_imgwidth(context) / 2 + x / 2 - (width * strlen(text)) / 2, y - height / 2, width, height, font, color, text);
	}
    }
}

static void (*control1) (int);
static int called = 0;

static void mycontrol(void *data, int i)
{
    called = 1;
    if (control1 != NULL)
	control1(i);
}
static void mycontrol2(void *data, int i)
{
   ((void (*) (int))data)(i);
}

int bbupdate()
{
    int ch;
    tl_update_time();
    TIME = tl_lookup_timer(scenetimer);
    tl_process_group (syncgroup, NULL);
    ch = aa_getkey(context, 0);
    switch (ch) {
    case 's':
    case 'S':
    case AA_BACKSPACE:
	finish_stuff = 1;
	break;
    case AA_ESC:
    case 'q':
	finish_stuff = 1, quitnow = 1;
    }
    return (ch);
}

void timestuff(int rate, void (*control) (int), void (*draw) (void), int maxtime)
{
    int waitmode = 0, t;
    tl_timer *timer;
    bbupdate();
    /*starttime = TIME; */
    endtime = starttime + maxtime;
    timer = tl_create_timer();
    if (control == NULL) {
	rate = -40;
    }
    if (rate < 0) {
	waitmode = 1, rate = -rate;
	control1 = control;
	tl_set_multihandler(timer, mycontrol, NULL);
    }
    else
	tl_set_multihandler(timer, mycontrol2, control);
    tl_set_interval(timer, 1000000 / rate);
    tl_add_timer(syncgroup, timer);
    tl_reset_timer(timer);
    tl_slowdown_timer (timer, starttime - TIME);
    if (control != NULL)
	control(1);
    while (!finish_stuff && TIME < endtime) {
	called = 0;
	bbupdate();
	t = tl_process_group(syncgroup, NULL);
	if (TIME > endtime)
	    break;
	if (!called && waitmode)
	    tl_sleep(t);
	else {
	    if (draw != NULL)
		draw();
	}
    }
    starttime = endtime;
    tl_free_timer(timer);
}

void
bbwait (int maxtime)
{
  int wait;
  if (finish_stuff)
    return;
  bbupdate ();
  endtime = starttime + maxtime;

  wait = endtime - TIME;
  while (wait > 0)
    {
      int t;
      bbupdate ();
      t = tl_process_group (syncgroup, NULL);
      wait = endtime - TIME;
      if (wait < t)
	t = wait;
      tl_sleep (t);
    }
  starttime = endtime;
}

void bbflushwait(int maxtime)
{
    int wait;
    if (finish_stuff)
	return;
    bbupdate();
    wait = maxtime + starttime - TIME;
    if (wait > 0) {
	aa_flush(context);
    }
    bbwait(maxtime);
}

static int stage = 1;

int bbinit(int argc, char **argv)
{
    aa_defparams.supported|= AA_NORMAL_MASK | AA_BOLD_MASK | AA_DIM_MASK;
    aa_parseoptions(NULL, NULL, &argc, argv);
    if (argc != 1 && (argc != 2 || ((argv[1][0] <= '0' || argv[1][0] > '8') && strcmp(argv[1], "-loop")))) {
	printf("Usage: bb [aaoptions] [number]\n\n");
	printf("Options:\n"
	       "  -loop          play demo in infinite loop\n\n"
	       "AAlib options:\n%s\n", aa_help);
	exit(1);
    }
    context = aa_autoinit(&aa_defparams);
    if (!context) {
	printf("Failed to initialize aalib\n");
	exit(2);
    }
    if (!aa_autoinitkbd(context, 0)) {
	aa_close(context);
	printf("Failed to initialize keyboard\n");
	exit(3);
    }
    if (argc == 2 && !strcmp(argv[1], "-loop"))
	loopmode = 1;
    else if (argc == 2)
	stage = atol(argv[1]);
    aa_hidecursor(context);
    return 1;
}

int bb(void)
{
    aa_gotoxy(context, 0, 0);
    introscreen();
    params = aa_getrenderparams();
    aa_render(context, params, 0, 0, 1, 1);
    font = uncompressfont( /*context->params.font */ &aa_font16);
    scenetimer = tl_create_timer();
    srand(time(NULL));
    if (stage != 1)
	finish_stuff = 1;
    do
	switch (stage) {
	default:
	case 1:
	    load_song("bb.s3m");
	    bbupdate();
	    starttime = endtime = TIME;

	    aa_resize (context);
	    scene1();
	    aa_resize (context);
	    scene3();
	    if (quitnow)
		goto quit;
	    aa_resize (context);
	    vezen(&fk1, &fk2, &fk3, &fk4);
	    messager("FILIP KUPSA known as FK, Tingle Notions, Dawn Music\n"
		"birth: June 22 1979, Tabor, Czech Republic, sex: male\n"
		     "\n"
	     "1992 - Changed his piano for 386/mp.com/pc-speaker music\n"
		     "1993 - Got his first Sound Blaster\n"
		     "1995 - Changed his SB for a new GUS technology\n"
		     "1996 - Composed his first great hits\n"
		     "1996 - FAT recomposition made by Windows 95\n"
		     "1997 - Released his musac in BB\n"
		     "\n"
		     "1998 - Got retired\n"
		     "\n"
		     "Contact address: via KT");
	    devezen2();
	    aa_resize (context);
	    scene4();
	    aa_resize (context);
	    scene2();
	    aa_resize (context);
	    if (quitnow)
		goto quit;
	    vezen(&ms1, &ms2, &ms3, &ms4);
	    messager("MOJMIR SVOBODA known as MS, TiTania, MSS, Bill\n"
		     "birth: ??, Tabor, Czech Republic, sex: ? male ?\n"
		     "\n"
		     "1993 - Installed Linux on his 386sx/25 + 40MB HDD\n"
		     "1994 - Removed Linux to make space for Doom\n"
		   "1995 - Reinstalled Linux on his 486Dx4/120 + 850MB\n"
		     "1996 - Removed Linux to make space for Windows 95\n"
		     "\n"
		     "1997 - Removed Windows 95 to make space for aalib\n"
		     "\n"
		     "Contact address: titania@mbox.vol.cz");
	    devezen3();
	    aa_resize (context);
	    scene8();
	    aa_resize (context);
	    scene6();
	    aa_resize (context);
	case 2:
	    if (quitnow)
		goto quit;
	    aa_resize (context);
	    vezen(&kt1, &kt2, &kt3, &kt4);
	    messager("KAMIL TOMAN known as KT, Kato, Whale, Bart\n"
		 "birth: May 19 1979, Tabor, Czech Republic, sex: male\n"
		     "\n"
		     "1993 - Became a linux extremist\n"
		     "1993 - Successful attempt to establish a secret organization\n"
		     "       Commandline Brotherhood\n"
		     "1995 - Action 'koules' - a secret project to train brotherhood\n"
		     "       members - covered under a game design\n"
		     "\n"
		 "1998 - Heading a new wave of command line revolution\n"
		     "\n"
		     "Contact address: toman@horac.ta.jcu.cz");
	    bbupdate();
	    starttime = endtime = TIME;
	    devezen1();
	    aa_resize (context);
	    if (quitnow)
		goto quit;
	    aa_resize (context);
	    scene7();
	    if (quitnow)
		goto quit;
	    aa_resize (context);
	    scene5();
	    if (quitnow)
		goto quit;
	    aa_resize (context);
	    scene10();
	    vezen(&hh1, &hh2, &hh3, &hh4);
	    messager("JAN HUBICKA known as HH, Jahusoft, HuJaSoft, JHS, UNIX, Honza\n"
		  "birth: Apr 1 1978, Tabor, Czech Republic, sex: male\n"
		     "\n"
		     "1991 - Installed underground hackers OS Linux\n"
		     "1995 - Headed Action 'koules'\n"
		     "1996 - Famous troan XaoS to convert all windows instalations\n"
		     "       into Linux\n"
		     "\n"
		     "1998 - Secret plan to make `Text Windows` system to confuse users\n"
		 "2001 - Planning an assassination of dictator Bill G.\n"
		     "\n"
		     "Contact address: hubicka@paru.cas.cz");
	    aa_resize (context);
	    devezen4();
	    if (quitnow)
		goto quit;
	    aa_resize (context);
	    credits();
	    if (quitnow)
		goto quit;
	case 3:
	    if (loopmode)
		break;
	    aa_resize (context);
	    credits2();
	}
    while (loopmode);
  quit:;
    aa_close(context);
    return (0);
}
